src/poetry/json/schemas/poetry.json

```diff
@@ -20,8 +20,7 @@
       "type": "object",
       "additionalProperties": false,
       "required": [
-        "name",
-        "url"
+        "name"
```

Comment:
I also tried some fancy `anyOf` `not` `pattern` magic, which worked correctly, but always gave the following message for invalid combinations: `[source.0] {'name': 'x'} is not valid under any of the given schemas`

Thus, I decided to be permissive here and check for invalid combinations in `create_package_source()` in `factory.py`. This way, we get way better error messages if there is a PyPI with a url or another source without a url.